piqueserver/auth.py

```diff
@@ -1,13 +1,77 @@
+from typing import Tuple, Dict, List
+from collections import defaultdict
 import abc
+from piqueserver.config import config, _Option
+login_retries = config.option('login_retries', 3)
+
+Details = Tuple[str, str]  # username, password
+
+
+class AuthError(Exception):
+    pass
 
 
 class BaseAuthBackend(abc.ABC):
     @abc.abstractmethod
-    def login(self, details: Tuple):
+    def login(self, details: Details) -> str:
+        """Verifies details and returns an user_role.
+        Raises AuthError if details are incorrect."""
+        pass
+
+    @abc.abstractmethod
+    def has_permission(self, connection, action: str) -> bool:
+        """Checks if a player has permission to perfom specific action."""
         pass
 
+    @abc.abstractmethod
+    def get_rights(self, user_type: str) -> List[str]:
+        """Returns list of actions an user type can perform"""
+        pass
+
+    def set_user_type(self, connection, user_type: str) -> None:
+        if user_type == 'admin':
+            connection.admin = True
+            connection.speedhack_detect = False
+        connection.user_types.add(user_type)
+        rights = set(self.get_rights(user_type))
+        connection.rights.update(rights)
+
+
+def notify_login(connection, user_type: str) -> None:
+    connection.on_user_login(user_type, True)
+    message = '{} logged in as ' + user_type
+    connection.send_chat(message.format('You'))
+    connection.protocol.irc_say(message.format('* ' + connection.name))
+
+_login_retries = defaultdict(login_retries.get) # type: Dict[str, int]
+
+def retries_left(connection) -> int:
+    return _login_retries[connection.address[0]]
+
+def log_failed_attempt(connection):
+    _login_retries[connection.address[0]] -= 1
+
 class ConfigAuthBackend(BaseAuthBackend):
-    """Auth backend that uses the [passwords] section of the connfig for
+    """Auth backend that uses the [passwords] section of the config for
     authentication"""
-    def login(self, username):
-        pass
+
+    def __init__(self):
+        self.passwords = config.option('passwords', default={}) # type: _Option
+        self.rights = config.option('rights', default={}) # type: _Option
+
+    def login(self, details: Details) -> str:
```

Comment:
These functions must return a result immediately, any remote queries done here would hold up the whole event loop. They should be `async def` or return `Deferred` instead.